cpp/src/arrow/util/async_generator.h

```diff
@@ -772,20 +782,26 @@ class ReadaheadGenerator {
   Future<T> operator()() {
     if (state_->readahead_queue.empty()) {
       // This is the first request, let's pump the underlying queue
-      state_->num_running.store(state_->max_readahead);
+      {
+        auto guard = state_->mutex.Lock();
+        state_->num_running = state_->max_readahead;
+      }
       for (int i = 0; i < state_->max_readahead; i++) {
         auto next = state_->source_generator();
         auto next_after_check = AddMarkFinishedContinuation(std::move(next));
         state_->readahead_queue.push(std::move(next_after_check));
       }
     }
     // Pop one and add one
-    auto result = state_->readahead_queue.front();
+    auto result = std::move(state_->readahead_queue.front());
     state_->readahead_queue.pop();
-    if (state_->finished.load()) {
+    auto guard = state_->mutex.Lock();
+    if (state_->finished) {
+      guard.Unlock();
       state_->readahead_queue.push(AsyncGeneratorEnd<T>());
     } else {
-      state_->num_running.fetch_add(1);
+      ++state_->num_running;
+      guard.Unlock();
```

Comment:
Maybe we can make the locking more compact and implicit (explicit unlocking in multiple branches could be somehow error-prone?), like:
```C++
bool finished = false;
{
  auto guard = state_->mutex.Lock();
  if (finished = state_->finished; !finished) {
    ++state_->num_running;
  }
}
if (finished) {
  // ...
} else {
  // ...
}
```

Reply:
done